src/main/java/bio/terra/janitor/service/cleanup/FlightManager.java

```diff
@@ -0,0 +1,113 @@
+    this.stairway = stairway;
+    this.janitorDao = janitorDao;
+    this.cleanupFlightFactory = cleanupFlightFactory;
+  }
+
+  /**
+   * Schedule a single resource for cleaning. Returns whether a flight id if a flight was attempted
+   * to be submitted to Stairway.
+   */
+  public Optional<String> submitFlight(Instant expiredBy) {
+    String flightId = stairway.createFlightId();
+    Optional<TrackedResource> resource = janitorDao.updateResourceForCleaning(expiredBy, flightId);
+    if (!resource.isPresent()) {
+      // No resource to schedule.
+      return Optional.empty();
+    }
+    submitToStairway(flightId, resource.get());
+    return Optional.of(flightId);
+  }
+
+  /**
+   * Recover tracked resources with flight ids in the Janitor's storage that are not known to
+   * Stairway. Resubmit the flights, returning how many flights were resubmitted.
+   *
+   * <p>This function assumes that it is not running concurrently with any other submissions to
+   * Stairway, e.g {@link #submitFlight(Instant)}.
+   */
+  public int recoverUnsubmittedFlights() {
+    List<JanitorDao.TrackedResourceAndFlight> resourceAndFlights =
+        janitorDao.retrieveResourcesWith(CleanupFlightState.INITIATING, RECOVERY_LIMIT);
+    if (resourceAndFlights.size() == RECOVERY_LIMIT) {
+      logger.error(
+          "Recovering as many flights as the limit {}. Some flights may still need recovering.",
+          RECOVERY_LIMIT);
+    }
+    int submissions = 0;
+    for (JanitorDao.TrackedResourceAndFlight resourceAndFlight : resourceAndFlights) {
+      String flightId = resourceAndFlight.cleanupFlight().flightId();
+      try {
+        // If there is some flight state, then the flight has been submitted successfully and does
+        // not need to be recovered. There's nothing to do but wait for the flight to update it's
+        // state from INITIATING.
+        // TODO(wchamber): add metrics.
+        stairway.getFlightState(flightId);
+      } catch (FlightNotFoundException e) {
+        // Stairway does not know about the flightId, so we must not have submitted successfully.
+        // Try to resubmit.
+        submitToStairway(flightId, resourceAndFlight.trackedResource());
+        ++submissions;
+      } catch (DatabaseOperationException | InterruptedException e) {
+        logger.error(String.format("Error recovering flight id [%s]", flightId), e);
+      }
+    }
+    return submissions;
+  }
+
+  /** Submits a cleanup flight for the resource to Stairway, or fails logging any exceptions. */
+  private void submitToStairway(String flightId, TrackedResource resource) {
+    FlightFactory.FlightSubmission flightSubmission =
+        cleanupFlightFactory.createSubmission(resource);
+    try {
+      stairway.submitToQueue(
+          flightId, flightSubmission.clazz(), flightSubmission.inputParameters());
+    } catch (StairwayException | InterruptedException e) {
```

Comment:
Seems you just catch the exception and log it. Then you still increase the submissions by one in the caller method. Do we want to return boolean success, and don't do the increase if `success==false?`

Reply:
done